### PR TITLE
Add API to assemble CPU shards to a sharded tensor

### PR DESCRIPTION
**Pull request by jonb377**
Currently, to convert CPU shards to an XLAShardedTensor, the API `XLAShardedTensor::load_local_shards_` must be used. This loads CPU shards in-place to an existing sharded tensor. A more convenient method for use cases outside of distributed checkpointing is to directly assemble the shards into a global tensor on device.

This PR adds a private _XLAC API `_get_global_tensor_from_cpu_shards` to allow directly creating a sharded tensor from a list of CPU shards and an OpSharding generated by `Mesh::get_op_sharding`.

I currently plan to use this new API in a few places:
 - tpu.py::discover_master_worker_ip_address: place each host's IP into a global tensor and pull out the zeroth entry for worker 0. The existing IP discovery API currently doesn't work with SPMD, but it's needed for distributed checkpointing.
 - Distributed data loading: A per-device, sharding-aware dataloader can use this API to efficiently create device data from the loaded CPU shards.

### CODE REVIEW

File: torch_xla/runtime.py [hunk 1]
```diff
@@ -131,6 +131,12 @@ def global_device_count() -> int:
   return len(torch_xla._XLAC._xla_get_all_devices())
 
 
+@requires_pjrt
+def local_runtime_devices() -> List[str]:
```
Comment (jonb377) [inline torch_xla/runtime.py:135]: @JackCaoG What do you think about this API? I wanted to include it so there's some way to discover the local device ID ordering.

File: test/spmd/test_xla_sharding.py [hunk 2]
```diff
@@ -866,6 +866,71 @@ def test_op_sharding_cache(self):
     xs.mark_sharding(v, mesh, (0, None))
     self.assertEqual(met.counter_value("CreateOpSharding"), 2)
 
+  def test_from_cpu_shards(self):
+    # Create an OpSharding with all devices on a single axis
+    mesh = self._get_mesh((self.n_devices,))
+    op_sharding = mesh.get_op_sharding((0,))
```
Comment (yeounoh) [inline test/spmd/test_xla_sharding.py:872]: Can we break this down to 
```
partition_spec = (0,)
op_sharding = mesh.get_op_sharding(partition_spec)
```
for readability?

File: test/spmd/test_xla_sharding.py [hunk 3]
```diff
@@ -866,6 +866,71 @@ def test_op_sharding_cache(self):
...
+    shards = [torch.LongTensor([i]) for i in range(self.n_devices)]
+    global_tensor = XLAShardedTensor.from_cpu_shards(shards, op_sharding)
+    self.assertTrue(
+        torch.allclose(global_tensor.cpu(), torch.arange(self.n_devices)))
+
+    # Test incorrect number of shards
+    with self.assertRaises(RuntimeError):
+      XLAShardedTensor.from_cpu_shards(shards[:-1], op_sharding)
+
+    # Test replicated sharding. The result should equal a shard.
```
Comment (yeounoh) [inline test/spmd/test_xla_sharding.py:884]: nit, "equal to a shard." Handling replication transparently to the frontend is great. Thanks!

File: test/spmd/test_xla_sharding.py [hunk 4]
```diff
@@ -866,6 +866,71 @@ def test_op_sharding_cache(self):
...
+
+    # Test an invalid global shape - too many values.
+    with self.assertRaises(RuntimeError):
+      bad_size = torch.Size((2 * self.n_devices,))
+      XLAShardedTensor.from_cpu_shards(shards, op_sharding, bad_size)
+
+    if self.n_devices > 1:
+      # Test an invalid global shape - incorrect rank
+      with self.assertRaises(RuntimeError):
+        bad_size = torch.Size((self.n_devices // 2, 2))
```
Comment (yeounoh) [inline test/spmd/test_xla_sharding.py:898]: Does it work with 1 device case?

File: test/spmd/test_xla_sharding.py [hunk 5]
```diff
@@ -866,6 +866,71 @@ def test_op_sharding_cache(self):
...
+    self.assertTrue(
+        torch.allclose(global_tensor.cpu(), torch.arange(self.n_devices)))
+
+    # Test incorrect number of shards
+    with self.assertRaises(RuntimeError):
+      XLAShardedTensor.from_cpu_shards(shards[:-1], op_sharding)
+
+    # Test replicated sharding. The result should equal a shard.
+    op_sharding = mesh.get_op_sharding((None,))
+    shards = [torch.LongTensor([0])] * self.n_devices
```
Comment (yeounoh) [inline test/spmd/test_xla_sharding.py:886]: I think we should emphasize/comment that `from_cpu_shards` won't move the shards, but copy and the shards are detached.

File: test/spmd/test_xla_sharding.py [hunk 6]
```diff
@@ -866,6 +866,71 @@ def test_op_sharding_cache(self):
...
+      XLAShardedTensor.from_cpu_shards(shards, op_sharding, bad_size)
+
+    if self.n_devices > 1:
+      # Test an invalid global shape - incorrect rank
+      with self.assertRaises(RuntimeError):
+        bad_size = torch.Size((self.n_devices // 2, 2))
+        XLAShardedTensor.from_cpu_shards(shards, op_sharding, bad_size)
+
+      # Test a valid global shape - restrict the number of meaningful values
+      # to 1, treating the rest as padding.
```
Comment (yeounoh) [inline test/spmd/test_xla_sharding.py:902]: Is it always the case, or due to the replicated op sharding.

File: test/spmd/test_xla_sharding.py [hunk 7]
```diff
@@ -866,6 +866,71 @@ def test_op_sharding_cache(self):
...
+    with self.assertRaises(RuntimeError):
+      XLAShardedTensor.from_cpu_shards(shards[:-1], op_sharding)
+
+    # Test replicated sharding. The result should equal a shard.
+    op_sharding = mesh.get_op_sharding((None,))
+    shards = [torch.LongTensor([0])] * self.n_devices
+    global_tensor = XLAShardedTensor.from_cpu_shards(shards, op_sharding)
+    self.assertTrue(torch.allclose(global_tensor.cpu(), shards[0]))
+
+    # Test an invalid global shape - too many values.
```
Comment (yeounoh) [inline test/spmd/test_xla_sharding.py:890]: What happens if the global shape is smaller, do we treat the rest as padding or error?

File: test/spmd/test_xla_sharding.py [hunk 8]
```diff
@@ -866,6 +866,71 @@ def test_op_sharding_cache(self):
...
+
+      # Test a valid global shape - restrict the number of meaningful values
+      # to 1, treating the rest as padding.
+      global_shape = torch.Size((1,))
+      global_tensor = XLAShardedTensor.from_cpu_shards(shards, op_sharding,
+                                                       global_shape)
+      self.assertTrue(torch.allclose(global_tensor.cpu(), torch.arange(1)))
+
+      # Use a 2d mesh
+      mesh_2d = self._get_mesh((self.n_devices // 2, 2))
```
Comment (yeounoh) [inline test/spmd/test_xla_sharding.py:909]: Would this work with 1 device case?

File: test/spmd/test_xla_sharding.py [hunk 9]
```diff
@@ -866,6 +866,71 @@ def test_op_sharding_cache(self):
...
+      #  | TPU:4 | TPU:5 |     | 0 | 1 |
+      #  | TPU:6 | TPU:7 |     | 0 | 1 |
+      #
+      # Partial replication along the 0th axis represents a global tensor
+      # of torch.Tensor([[0, 1]]).
+      shards = [torch.LongTensor([[i]]) for i in range(2)] * (
+          self.n_devices // 2)
+      op_sharding = mesh_2d.get_op_sharding((None, 1))
+      global_tensor = XLAShardedTensor.from_cpu_shards(shards, op_sharding)
+      expected = torch.arange(self.n_devices)
```
Comment (yeounoh) [inline test/spmd/test_xla_sharding.py:929]: this is not used, but `torch.arange(2).reshape(1, 2)`?

File: torch_xla/csrc/init_python_bindings.cpp [hunk 10]
```diff
@@ -1660,6 +1660,66 @@ void InitXlaModuleBindings(py::module m) {
...
+            << "Please enable SPMD via `torch_xla.runtime.use_spmd()`";
+        auto local_devices = runtime::GetComputationClient()->GetLocalDevices();
+        XLA_CHECK(local_devices.size() == shards.size())
+            << "Must specify a shard for each local device";
+
+        if (!global_shape.has_value()) {
+          // Set a default value for the global shape based on the sharding
+          // type.
+          if (sharding.type() == xla::OpSharding::OTHER) {
+            // Infer the global shape to be the shard shape scaled by the tiling
```
Comment (yeounoh) [inline torch_xla/csrc/init_python_bindings.cpp:1681]: This may not hold true, in case of uneven tiling. Let's make a note on this.

Comment (yeounoh) [reply torch_xla/csrc/init_python_bindings.cpp:1681]: or, actually for the right most shard in each dim, we can add the sizes, as the padding is always on the last dims.

File: torch_xla/csrc/init_python_bindings.cpp [hunk 11]
```diff
@@ -1660,6 +1660,66 @@ void InitXlaModuleBindings(py::module m) {
...
+            // dimensionality.
+            auto tile_shape = sharding.tile_assignment_dimensions();
+            global_shape = std::vector<long int>();
+            for (int dim = 0; dim < shards[0].sizes().size(); ++dim) {
+              auto global_dim = tile_shape[dim] * shards[0].sizes()[dim];
+              global_shape->push_back(global_dim);
+            }
+          } else {
+            // If the sharding type is not tiled, the tensor shards are
```
Comment (yeounoh) [inline torch_xla/csrc/init_python_bindings.cpp:1690]: This may not hold true in the future, can you check for the replicated type, and have a catch-all block for raising error?

File: torch_xla/csrc/init_python_bindings.cpp (hunk 10 above)
Comment (yeounoh) [reply torch_xla/csrc/init_python_bindings.cpp:1681]: Ok we can't do that, since `"Input shard shape must include padding: " << shard.sizes()"`

File: torch_xla/experimental/xla_sharding.py [hunk 12]
```diff
@@ -87,6 +87,14 @@ def get_op_sharding(self,
     Return the OpSharding for the given partition spec. This is an expensive
     operation as the mesh grows, so the value is cached for reuse.
     """
+    partition_spec = _translate_named_partition_spec(self, partition_spec)
```
Comment (yeounoh) [inline torch_xla/experimental/xla_sharding.py:90]: Thanks for refactoring 👍

File: torch_xla/runtime.py (hunk 1 above)
Comment (yeounoh) [reply torch_xla/runtime.py:135]: I think this could be handy.

**Review by yeounoh — COMMENTED** [review]
Let some comments.

File: test/spmd/test_xla_sharding.py (hunk 4 above)
Comment (jonb377) [reply test/spmd/test_xla_sharding.py:898]: Good call, I can make this a single-device test.

File: test/spmd/test_xla_sharding.py (hunk 5 above)
Comment (jonb377) [reply test/spmd/test_xla_sharding.py:886]: That's a good point. Do you think adding that to the docstring is sufficient?

File: test/spmd/test_xla_sharding.py (hunk 6 above)
Comment (jonb377) [reply test/spmd/test_xla_sharding.py:902]: This is always the case, I'll add another test for tiled sharding.

File: test/spmd/test_xla_sharding.py (hunk 9 above)
Comment (jonb377) [reply test/spmd/test_xla_sharding.py:929]: Thanks! I'll update the assertion to use `expected`.

File: torch_xla/csrc/init_python_bindings.cpp (hunk 10 above)
Comment (jonb377) [reply torch_xla/csrc/init_python_bindings.cpp:1681]: Let me know if I should revisit this. I decided on this approach because padding can cross over multiple devices, e.g. sharding a tensor with shape `(1, 2)` on the mesh `(1, 4)` will have shards with shape `(1, 1)` with no real data on the last two devices.

File: test/spmd/test_xla_sharding.py (hunk 8 above)
Comment (jonb377) [reply test/spmd/test_xla_sharding.py:909]: Oh I see what you mean, by making the mesh `(1, self.n_devices)`? I wanted this test to have multiple tile dims to cover the inferred global shape. I can make these tests run single-device as well.

File: test/spmd/test_xla_sharding.py (hunk 7 above)
Comment (jonb377) [reply test/spmd/test_xla_sharding.py:890]: It will infer padding up to a point, but if it's too small it will also fail the assertion. I'll add another test case for that.

File: torch_xla/csrc/init_python_bindings.cpp [hunk 13]
```diff
@@ -1660,6 +1660,66 @@ void InitXlaModuleBindings(py::module m) {
...
           return std::nullopt;
         });
+  // Reassemble the CPU shards into a global tensor. A new sharded tensor is
+  // created from the local shards with the provided sharding annotation
+  // attached. The order of the shards should coincide with the order of
+  // devices returned by `torch_xla.runtime.local_runtime_devices()`.
+  m.def(
+      "_global_tensor_from_cpu_shards",
+      [](const std::vector<at::Tensor>& shards, const xla::OpSharding& sharding,
+         std::optional<std::vector<long int>>& global_shape) -> at::Tensor {
```
Comment (alanwaketan) [inline torch_xla/csrc/init_python_bindings.cpp:1670]: Can we use int64_t here?

File: torch_xla/csrc/init_python_bindings.cpp [hunk 14]
```diff
@@ -1660,6 +1660,66 @@ void InitXlaModuleBindings(py::module m) {
...
+              global_shape->push_back(global_dim);
+            }
+          } else {
+            // If the sharding type is not tiled, the tensor shards are
+            // replicated.
+            global_shape = shards[0].sizes().vec();
+          }
+        }
+
+        xla::Shape tensor_shape =
```
Comment (alanwaketan) [inline torch_xla/csrc/init_python_bindings.cpp:1696]: Is this supposed to be global shape? Then why use shards[0]?

File: torch_xla/experimental/xla_sharded_tensor.py [hunk 15]
```diff
@@ -102,6 +102,19 @@ def __new__(cls, elem: torch.Tensor, *args, **kwargs):
     r.global_tensor = elem.detach() if r.requires_grad else elem
     return r
 
+  @staticmethod
+  def from_cpu_shards(shards: List[torch.Tensor],
+                      sharding: torch_xla._XLAC.OpSharding,
+                      global_shape: torch.Size = None):
```
Comment (alanwaketan) [inline torch_xla/experimental/xla_sharded_tensor.py:108]: What's the benefit of providing a global_shape?

File: torch_xla/csrc/init_python_bindings.cpp (hunk 14 above)
Comment (jonb377) [reply torch_xla/csrc/init_python_bindings.cpp:1696]: This is a bit of a shortcut to use `CreateComputationShapeFromTensor`, the shape's dimensions are updated in the next few lines. I'll actually just switch to the lower-level API `MakeArrayShapeFromDimensions` instead, that will be easier to follow.

File: torch_xla/csrc/init_python_bindings.cpp (hunk 13 above)
Comment (jonb377) [reply torch_xla/csrc/init_python_bindings.cpp:1670]: Good point, I'll change this.

File: torch_xla/experimental/xla_sharded_tensor.py (hunk 15 above)
Comment (jonb377) [reply torch_xla/experimental/xla_sharded_tensor.py:108]: If the shards are padded, the global_shape can be used to ignore padding.

File: torch_xla/runtime.py (hunk 1 above)
Comment (jonb377) [reply torch_xla/runtime.py:135]: I'm removing this, we'll keep the new `_global_tensor_from_cpu_shards` private.

File: torch_xla/csrc/init_python_bindings.cpp (hunk 10 above)
Comment (alanwaketan) [reply torch_xla/csrc/init_python_bindings.cpp:1687]: ping @yeounoh

**Review by yeounoh — APPROVED** [review]
LGTM

File: torch_xla/csrc/init_python_bindings.cpp (hunk 10 above)
Comment (yeounoh) [reply torch_xla/csrc/init_python_bindings.cpp:1687]: I think for the most part, user should specify the global shape. I am thinking that we should actually make it explicit, not optional... but if not, this way of handling, inferring the default shape is good.

Comment (jonb377) [reply torch_xla/csrc/init_python_bindings.cpp:1687]: That's a good point... I chose to let it be inferred for convenience, since for e.g. distributed data loading, the shards are provided on CPU with the correct padded local shape, and deriving the global shape could be difficult for more sophisticated shardings.

I'll go ahead and land with it optional for now, since we're keeping the API private. Thanks Yeounoh!

**Comment by jonb377** [comment]
Thanks @yeounoh and @alanwaketan for the review. I'll merge after a TPU CI run.

**Review by alanwaketan — APPROVED** [review]
LGTM.
